extensions/notebook/src/jupyter/jupyterServerInstallation.ts

```diff
@@ -165,6 +165,14 @@ export class JupyterServerInstallation implements IJupyterServerInstallation {
 			let pythonExists = await utils.exists(this._pythonExecutable);
 			if (!pythonExists || forceInstall) {
 				await this.installPythonPackage(backgroundOperation, this._usingExistingPython, this._pythonInstallationPath, this.outputChannel);
+				// reinstall pip to make sure !pip command works
+				if (!this._usingExistingPython) {
+					let versionCmd = `"${this._pythonExecutable}" -m pip --version`;
+					let versionCmdResult = await this.executeBufferedCommand(versionCmd);
+					let version = versionCmdResult.substring(4, 10);
```

Comment:
Version numbers aren't guaranteed to all have the same length. If the version was something like "20.2" then by the 10th index you've already started pulling in other characters from the version message. Given a message that looks like this: "pip 20.2 from ...", substring(4,10) would give us "20.2 f". Then you can get the opposite problem if the version number is too long, like if it includes some kind of pre-release tag.

Reply:
Looking at the pip release history, we have version "21.0" which hits the too short issue, and "10.0.0b2" which hits the too long issue.

Reply:
You can get the installed list of packages using the getInstalledPipPackages() method, which has all the versions listed. Frustratingly, "pip show" doesn't have a json output format like "pip list", so you'd have to get the list of all the packages.